stacks/spica/src/function/pages/add/add.component.ts

```diff
@@ -178,4 +180,21 @@ export class AddComponent implements OnInit, OnDestroy {
         this.dependencyInstallPending = false;
       });
   }
+
+  checkHandlers() {
+    this.isHandlerDuplicated = false;
+    this.function.triggers.forEach(trigger => {
+      const duplicatedHandler = this.function.triggers.filter(
+        item => item.handler == trigger.handler
+      );
+      if (duplicatedHandler.length > 1) {
+        this.isHandlerDuplicated = true;
+      }
+    });
+  }
```

Comment:
@htuna07 Shared a shorter version of the method. 

```
checkHandlers() {
     this.function.triggers.forEach(trigger => {
       const duplicatedHandler = this.function.triggers.filter(
         item => item.handler == trigger.handler
       );
       this.isHandlerDuplicated = duplicatedHandler.length > 1;
       }
     });
   }
```

Another thing is changing the method so it doesn't modify the state of the class but return the result as a boolean would make it more testable as you wouldn't need setting the `function` variable up of the class.

Basically passing a list as parameter and turning a boolean value would make this function a pure one which is more secure.

Reply:
This logic will set the isHandlerDuplicated true when the last value of an array isn't duplicated, but previous values could be duplicated. 

For example ;
The array that will be checked is `["value1","value1","value2"]`. This logic will set isHandlerDuplicated value false for the first and second items, but the last item will set this value true. 

Also, every change in the handler input fields, this function will be called and set the isHandlerDuplicated value. When the isHandlerDuplicated value changes, the error which is dependent on this boolean value will detect these changes and show an error or not. That's why we store this boolean value as global.

Reply:
@htuna07 I see. Is it possible using https://angular.io/guide/form-validation here?

Reply:
@htuna07 Have you seen the code change proposed in my first comment? It removes some lines without changing the usage of the function.

Reply:
We use the template-driven validation instead of reactive form validation. I tried to use the same validation way. In the template-driven forms, we need to create a validator which is checks this duplication but creating a new validator will cause us to spend more time on this basic validation.

Reply:
I see, thanks for the information.